lib/nerves_hub/deployments.ex

```diff
@@ -82,19 +83,64 @@ defmodule NervesHub.Deployments do
     end
   end
 
+  @doc """
+  Update a deployment
+
+  Updating a deployment is a big task. Devices will be notified of the change when:
+  - Firmware changes, all devices will be told of the new firmware to update
+  - Conditions change, all devices will have the deployment removed and told about the
+    change, any devices that don't have a deployment and are online will be told about
+    the conditions changing to check for a deployment again
+  - If now active, any devices without a deployment will be told to reevaluate
+  - If now inactive, devices will have the deployment removed and told about the change
+  """
   @spec update_deployment(Deployment.t(), map) :: {:ok, Deployment.t()} | {:error, Changeset.t()}
   def update_deployment(deployment, params) do
-    deployment
-    |> Deployment.with_firmware()
-    |> Deployment.changeset(params)
-    |> Repo.update()
-    |> Repo.reload_assoc(:firmware)
-    |> case do
+    changeset =
+      deployment
+      |> Deployment.with_firmware()
+      |> Deployment.changeset(params)
+
+    case Repo.update(changeset) do
       {:ok, deployment} ->
+        deployment = Repo.preload(deployment, [:firmware], force: true)
+
+        # if the conditions changed, we should reset all devices and tell any connected
+        if Map.has_key?(changeset.changes, :conditions) do
+          Device
+          |> where([d], d.deployment_id == ^deployment.id)
+          |> Repo.update_all(set: [deployment_id: nil])
+
+          broadcast(deployment, "deployments/conditions-changed", deployment.conditions)
+          broadcast(:none, "deployments/conditions-changed", deployment.conditions)
+
+          description = "deployment #{deployment.name} conditions changed and removed all devices"
+          AuditLogs.audit!(deployment, deployment, :update, description)
+        end
+
+        # if is_active is false, wipe it out like above
+        # if its now true, tell the none deployment devices
+        if Map.has_key?(changeset.changes, :is_active) do
+          if deployment.is_active do
+            broadcast(:none, "deployments/is-active-changed")
+          else
+            Device
+            |> where([d], d.deployment_id == ^deployment.id)
+            |> Repo.update_all(set: [deployment_id: nil])
+
+            broadcast(deployment, "deployments/is-active-changed")
+
+            description = "deployment #{deployment.name} is inactive and removed all devices"
+            AuditLogs.audit!(deployment, deployment, :update, description)
+          end
+        end
+
         fetch_and_update_relevant_devices(deployment)
```

Comment:
```suggestion
        broadcast(deployment, "deployments/update")
```

Seems like we could just replace these function calls with the broadcast since that is all that is happening now (no fetching and things)

Reply:
One extra thought - Are we ever going to have a race here between publishing the messages above which alter the deployment linked in the device socket and this broadcast telling the socket to prep and send an update message?

Reply:
`fetch_and_update_relevant_devices/1` is used somewhere, I don't remember exactly but getting rid of it caused a compilation warning so I left it. I can update things though.